### PR TITLE
Add support for PROXY protocol

### PR DESCRIPTION
**Pull request by Sir-Photch**
Suggested in #28.
Right now, receiving v1 works, but the information is just dumped into the command line. Should I put the proxy-protocol information into `struct client`?

### CODE REVIEW

**Comment by omar-polo** [comment]
Hello,

thanks for the the effort and sorry for the delay!

I haven't tried the diff yet, but it looks fine to me.  I have a few nits before we can merge this, but first I'd ask you if it would be possible amend the indentation changes in `server.c` in the first commit.  It makes the diff harder to review.

Thanks!

**Comment by Sir-Photch** [comment]
> but first I'd ask you if it would be possible amend the indentation changes in server.c in the first commit.

Yes, sorry. Apparently I told my editor to auto format at some point; I will take care of it.

We should probably add some unit-tests for parsing. And what do you think about fuzzing? Clang supports this AFAIK, but I am not sure about gcc.
Also, is this parsing happening in a process that drops its privileges? (It was (is) a little daunting to write network facing C code :b)

**Comment by omar-polo** [comment]
> Yes, sorry. Apparently I told my editor to auto format at some point; I will take care of that.

no problem.

> We should probably add some unit-tests for parsing.

Yep, I'd really like some tests for this.  We have both unit tests (although only used for the IRI parsing code) and runtime tests.  If we had support for the proxy protocol also for, ehm, proxying connections it would be easy to do.

But no rush for this, we'll get to it after the diff is in a state where it can be merged :)

> And what do you think about fuzzing? Clang supports this AFAIK, but I am not sure about gcc.

I wouldn't mind having something for fuzzying.  Maybe this is the chance I needed to look more into it.  AFAIK the compiler doesn't really matter, but you can use stuff like asan and ubsan too which may depend on the compiler/platform.

> Also, is this parsing happening in a process that drops its privileges? (It was (is) a little daunting to write network facing C code :b)

Yes, the connection handling runs in a process that has dropped privileges, chrooted and, depending on the system, sandboxed itself.

> If you want to test the behaviour with nginx in front, you can refer to this snippet:

Thanks!  It'll take be a bit to be able to actually test this, but many thanks for the nginx snippet!

**Comment by Sir-Photch** [comment]
Hi, I've cleaned up the whitespace changes. My actual changes should still be there. :)
`make regress` runs fine indicating that non-proxy-protocol connections are fine. (But there are no test for the protocol yet, of course.)

**Review by omar-polo — CHANGES_REQUESTED** [review]
I've left a first round of comments/nits.  Once these are fixed I believe we can merge this and start talking about the follow-ups.

There's no need to do everything in this PR, but after it's merged I think it would be a good idea to parse the protocol only if specified in the config.  (maybe I'd add an optional keyword to be used in the listen directive such as `listen on * port 1965 proxy-v1`)  Otherwise "normal" clients could spoof their IP address.

Having some support for the proxy protocol when proxying requests would also be nice, so that we could write some tests for it.  But again, no need to do all of this in this PR, we can work on that as a follow-up.

If you have any doubt or question please don't refrain from asking!

Thanks!

File: proxy-proto.c
Comment (omar-polo) [reply proxy-proto.c:192]: Wooops, this one didn't get sorted as I expected.  You'll find below the suggestion I'm mentioning here, sorry.

File: buflayer.c
Comment (Sir-Photch) [reply buflayer.c:14]: Originally this was for me to test if the parsing works with an incomplete proxy-protocol line

File: proxy-proto.c
Comment (Sir-Photch) [reply proxy-proto.c:31]: What if the first data packet we receive is an incomplete line of the header? Is it still valid to alter the buffer?

File: gmid.h
Comment (Sir-Photch) [reply gmid.h:302]: Sure, I just thought that this "growable buffer" is nothing specific to its contents. And, who doesn't like powers of two? :b

File: proxy-proto.c
Comment (Sir-Photch) [reply proxy-proto.c:47]: My thought process was:

1. keep on reading until end of buffer
2. if everything we read is valid parts of the protocol, return `TLS_WANT_POLLIN`
3. if at some point we see something funny, pass the unaltered buffer to libtls
4. if we reach `\r\n`, cut off everything else and pass it to libtls while parsing the rest

File: proxy-proto.c
Comment (Sir-Photch) [reply proxy-proto.c:193]: I don't understand what you mean by your second comment here? 😅

File: proxy-proto.c
Comment (Sir-Photch) [reply proxy-proto.c:160]: Again, this was considering incomplete headers.
Those are some rudimentary checks to see if the chars we have available can even be considered as a complete ipv4 address. Mostly, line 152 and 153 see if we should keep reading. Line 155f may be over top though, yes.

File: server.c
Comment (Sir-Photch) [reply server.c:1351]: Hm, but what about `"PROXY UNKNOWN\r\n"`?
Also, do I see correctly that `rhost` (IP) and `rserv` (port) are supposed to be strings? But why are they set to be 1025 (32) long? (For resolved hosts?)

Comment (Sir-Photch) [reply server.c:1351]: Also, by numeric you mean `2130706433` for `127.0.0.1`?

File: proxy-proto.c
Comment (omar-polo) [reply proxy-proto.c:193]: I meant that I expected the comments to appear in a different order.  I wanted this "chunk" to be after the next one. ^^"

File: proxy-proto.c
Comment (omar-polo) [reply proxy-proto.c:47]: Ah yes, I forgot to suggest one more thing.  To make parsing easier, we could keep reading until we find CRLF (or fill the buffer, in that case we abort the connection.)

At that point, you have the whole protocol line in memory and parsing becomes easier.  That, together with my other suggestion about changing the underlying string, should make the code easier to follow I believe.

(the rationale is that the proxy protocol is thought to fit one TCP packet so in practice we shouldn't issue more than one read(2), and even if we have it's just 108 bytes at maximum.)

File: server.c
Comment (omar-polo) [reply server.c:1350]: nope, sorr,y for numeric I meant to have rhost "127.0.0.1" and rserv "1965" (for e.g.)

In the "PROXY UNKNOWN" case I don't think we have much to do.  Maybe add a log_debug (but could be very well be a log_info or log_warnx even -- it's not expected to have the "UNKNOWN" protocol) and leave `rhost` and `rserv` as is.

regarding your second question, well, the limits are, unfortunately, arbitrary.  32 is just an upper bound (that I should probably reduce to 8 at least...).

File: proxy-proto.c
Comment (Sir-Photch) [reply proxy-proto.c:47]: Okay, seems reasonable.
Strictly speaking though, the v1 spec says:
> For "UNKNOWN", the rest of the line before the CRLF may be omitted by the sender, and the receiver must ignore anything presented before the CRLF is found.

Which I read as "there does not necessarily need to be a limit to header length for UNKNOWN protocols". But I suppose since there are no "unknown" protocols (yet?) it is fine to ignore that?

Comment (omar-polo) [reply proxy-proto.c:47]: That's a good point.  The specification says that

> So a 108-byte buffer is always enough to store all the line and a trailing zero for string processing.

but only after showing the maximum line for a non-UNKNOWN protocol, so I can see why there's space for different interpretations.

In any case, I don't think we should go that much out of our way to handle an hypothetical UNKNOWN line with more than 107 bytes of payload, also because I don't really like the idea of keep reading indefinitely for an edge case (I believe) noone is using in practice.

So let's just assume that also UNKNOWN protocols fit the 108 byte buffer.  If we fill the buffer without seeing the CRLF, let's just drop the connection with a nice error in the log and move on.

**Comment by Sir-Photch** [comment]
I rewrote the parsing according to your suggestions. Test are obviously failing now since proxy-protocol is expected now, what's missing is some kind of configuration for `server_accept()` to set `c->buf.should_buffer` accordingly.

File: server.c
Comment (omar-polo) [inline server.c:1332]: nit: setting errno is not needed.  read will set it if it fails, so no need to explicitly zero it.

File: server.c
Comment (omar-polo) [inline server.c:1355]: strncpy(3) is unfortunately one of those strange functions it's hard to use and are only seldomly useful.  TLDR; in this case it should be `strlcpy(c->host, "UNKNOWN", sizeof(c->host))`.

(str**N**cpy takes the maximum length of the second argument, not the first.  in this case there's plenty of space so in practice it will never fail.  Still, it's the wrong function)

File: ge.c
Comment (omar-polo) [inline ge.c:274]: why this change?

**Review by omar-polo — COMMENTED** [review]
This starts to look good, thank you for taking the time to improve it!

I'll push a commit to enable the proxy proto via a configuration knob.

File: server.c
Comment (omar-polo) [inline server.c:1347]: missed one thing: `close(c->fd)` here is a bit dangerous because later in `client_close` we'll close it again.  file descriptor numbers can be reused, so we might end up closing another client socket!

So, here we should IMHO log something (e.g. `log_warnx("invalid proxy protocol header")` or something), then return -1.  `tls_read()` will then bubble the error up to libevent which will call `client_error` where we'll finally properly close it.

**Comment by omar-polo** [comment]
To be fair it's a bit lame the way I've added the `proxy-v1` keyword.  I'll refactor a bit later how `listen` is processed so it's tidier.

After the pending thing will be fixed I believe we can merge this!  I hope you won't take it bad, but I'll probably rebase it against the master (I prefer to have a linear history and avoid merges), squash some commits, and pass `proxy-proto.c` thru [knf](https://github.com/mptre/knfmt), mostly for indentation, for consistency with the rest of the code.

Thanks! :)

File: server.c
Comment (Sir-Photch) [reply server.c:1353]: `c->host` or `c->rhost`?

File: proxy-proto.c
Comment (Sir-Photch) [reply proxy-proto.c:90]: good spot!

File: ge.c
Comment (Sir-Photch) [reply ge.c:274]: I figured if `gemexp` is for experimenting, a verbose log output might be useful as well. This was because of the `log_debug()` when we parsed the proxy protocol.

I can revert this, no problem. I though that this might be something to consider, though.

**Comment by Sir-Photch** [comment]
> I hope you won't take it bad, but I'll probably rebase it against the master (I prefer to have a linear history and avoid merges), squash some commits, and pass `proxy-proto.c` thru [knf](https://github.com/mptre/knfmt), mostly for indentation, for consistency with the rest of the code.

No problem, this is your codebase after all, and I admit that my commit-messages are not necessarily very meaningful. (Also some of them were drafts, so no reason not to squash.)

Thanks for pointing me to knf though, it's very interesting how OpenBSD folk do things :)

File: server.c
Comment (omar-polo) [reply server.c]: I meant `c->rhost`, sorry.  `host` is the requested host from the user, `rhost` the remote host from where the client is connecting.  Actually the naming could be better, it's a bit confusing :)

Comment (omar-polo) [reply server.c]: also, regarding the different APIs: there's the very old `strcpy` and `strcat` which assume infinite storage and so they're usually a bad sign when you seem them.  `strncpy` and `strncat` are weird because they assume the string you're copying to has infinite space but allows you to specify the length of the string you're copying from.  There could be some niche use for these, but they're usually a mistake too when you see someone using it.

`strlcpy` and `strlcat` IMHO are saner since they allow you to specify the length of the storage you're copying to.

**Review by omar-polo — APPROVED** [review]

**Comment by omar-polo** [comment]
I"ve merged it manually, so unfortunately on github it will show as 'closed' instead of merged :(

Thank you for working on this and enduring my requests :)

There are two follow-up commits that I'd like to point out since they may be of interest for you and that I noticed only after starting to playing with this.

 - https://github.com/omar-polo/gmid/commit/acbe7a53131d0ba42bf919ea7a11dc7d1b60fb9c
 - https://github.com/omar-polo/gmid/commit/2a090b866d78d8e9a06a1e49e6b48cd4b5a488d0

**Comment by omar-polo** [comment]
Just FYI I've started to fuzz some parts, the proxy protocol and the IRI parser.  See `regress/fuzz/README.md` if you're interested :)

**Comment by omar-polo** [comment]
I think I've found an issue with the current implementation: we expect both addresses to be of the same family, i.e. `PROXY TCP6 ::1 127.0.0.1 30677 1965` is considered malformed, while I think it should be supported.  (proxying from one family to another is not uncommon)

I'll fix it since I'd like to release 2.1 soon, but wanted to let you know.

**Comment by Sir-Photch** [comment]
Ah, yes, good spot. I misread the spec and assumed that the INET family was for both addresses. 

Anyway, thanks for taking care about this! I'm excited that this now landed in 2.1 :)

**Pull request closed without merging**
